scripts/core/superdesk-api.d.ts

```diff
@@ -1,9 +1,41 @@
+/* tslint:disable */
+
 declare module 'superdesk-api' {
-    export interface IExtension {
-        activate: (superdesk: ISuperdesk) => void;
+    // TYPESCRIPT TYPES
+
+    export interface DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {};
+
+    export type DeepReadonlyObject<T> = {
+        readonly [P in keyof T]: DeepReadonly<T[P]>;
+    }
+
+    export type DeepReadonly<T> =
+        T extends Function
+            ? T
+            : T extends Array<infer U>
+                ? DeepReadonlyArray<U>
+                : DeepReadonlyObject<T>;
+
+    export type Omit<K, V> = Pick<K, Exclude<keyof K, V>>;
+
+
+
+    // EXTENSIONS
+
+    export interface IExtensionActivationResult {
```

Comment:
One of the main things I'm trying to achieve with the extensions is to minimize coupling between implementation details of the core and extensions API. That allows for greater flexibility in changing the implementation details while maintaining the same API.

Imagine I wanted an extension to be able to register menu items. A naive approach would be to give the extension access to the `ul` element, and it would insert DOM nodes to it. That would work, but would also cause some issues:
* screen jumping around / flickering while different extensions insert items at different times
* if we wanted to add sorting of menu items, it wouldn't be straight-forward since we would have DOM nodes, not text. We could try getting `innerText`, but extensions might be using an icon-font for adding icons in front of the text and that would mess up the sorting. On top of that, extensions would be able to insert items at any time which means we would have to listen to DOM changes and re-sort afterward which wouldn't even have been possible a few years ago without recently introduced mutation observer.
* if the menu is closed, an extension wouldn't know that and might be doing HTTP requests which would slow down the app unnecessarily.

The solution for such issues is to limit what the API exposes invert the control - that is to call the extensions ourselves instead of allowing extensions call us. Continuing with the menu example, instead of allowing an extension to insert DOM nodes, I want to allow it to provide a function the core would call when it wants to render the menu for example. The function would return a promise, the core would wait for all results render the menu. We are also not giving it access to the DOM node, but are forcing it to return text which allows us to control the rendering. I want extensions to only be able to provide such functions once, so I have extensions returning them from `activate` function and call it activation result.

Feel free to suggest better naming anywhere in the code.

Reply:
yeah I get your point on extensions but I was just confused by the naming, not a big deal tho, I didn't think of any better naming